wallets/server.js

```diff
@@ -11,120 +11,231 @@ import * as blink from 'wallets/blink/server'
+    descriptionHash = undefined
+  }
+
+  const wallets =
+   useWallet
+     ? [useWallet] // if a wallet is specified, use it
+     : await models.wallet.findMany({
+       where: { userId: receiverUserId, enabled: true },
+       orderBy: [
+         { priority: 'asc' },
+         // use id as tie breaker (older wallet first)
+         { id: 'asc' }
+       ]
+     })
+
+  let i = attempt
+  if (i >= wallets.length) i = 0
+  for (; i < wallets.length; i++) {
+    const wallet = wallets[i]
 
-  for (const wallet of wallets) {
     const w = walletDefs.find(w => w.walletType === wallet.type)
     try {
+      if (!w) throw new Error(`wallet type ${wallet.type} not found`)
       if (!canReceive({ def: w, config: wallet.wallet })) {
         continue
       }
-
       const { walletType, walletField, createInvoice } = w
 
       const walletFull = await models.wallet.findFirst({
         where: {
-          userId,
+          userId: receiverUserId,
           type: walletType
         },
         include: {
           [walletField]: true
         }
       })
-
-      if (!walletFull || !walletFull[walletField]) {
-        throw new Error(`no ${walletType} wallet found`)
-      }
-
-      // check for pending withdrawals
-      const pendingWithdrawals = await models.withdrawl.count({
-        where: {
-          walletId: walletFull.id,
-          status: null
-        }
-      })
-
-      // and pending forwards
-      const pendingForwards = await models.invoiceForward.count({
-        where: {
-          walletId: walletFull.id,
-          invoice: {
-            actionState: {
-              notIn: PAID_ACTION_TERMINAL_STATES
-            }
-          }
-        }
-      })
-
-      console.log('pending invoices', pendingWithdrawals + pendingForwards)
-      if (pendingWithdrawals + pendingForwards >= MAX_PENDING_INVOICES_PER_WALLET) {
-        throw new Error('wallet has too many pending invoices')
-      }
-      console.log('use wallet', walletType)
-
-      const invoice = await withTimeout(
-        createInvoice({
+      await checkWallet(models, walletFull, walletField, walletType)
+
+      console.log('use wallet', wallet.type)
+      if (hodl || wrapped) {
+        // we always wrap p2p invoices unless it is a direct non-hodl invoice
+        const wrappedMsats = msats
+        const innerMsats = BigInt(msats) * (100n - BigInt(sybilFeePercent ?? ZAP_SYBIL_FEE_PERCENT)) / 100n
+        assert(innerMsats <= wrappedMsats, 'innerMsats must be less than or equal to wrappedMsats')
+
+        const invoice = await withTimeout(
+          createInvoice({
+            msats: toPositiveNumber(innerMsats),
+            description,
+            descriptionHash,
+            expiry
+          }, walletFull[walletField]), 10_000)
+        await checkInvoice(models, wallet, invoice, innerMsats)
+        if (!lnd) throw new Error('no lnd provided for wrapping')
+        const { invoice: wrappedInvoice, maxFee } = await wrapInvoice(invoice,
```

Comment:
we do wrapping here, so if for whatever reason this wallet has some incompatibility with our wrapping code, we fallback to the next one